cpp/lib/src/RegexMatcher.cc

```diff
@@ -172,6 +172,62 @@ std::string ThreadSafeRegexMatcher::replaceAll(const std::string &subject, const
 }
 
 
+static std::string InsertReplacement(const ThreadSafeRegexMatcher::MatchResult &result, const std::string &replacement_pattern) {
```

Comment:
It appears that this function, apart from the name on line 187 and the type of the first parameter is identical to the other InsertReplacement function in this file.  This calls for replacing both functions with a function template.  Just call the type of the first argument something like MatchedGroups and the name of the argument matched_groups and you should be able to get by with a single template function.